onnxscript/function_libs/torch_lib/new_graph_prototype.py

```diff
@@ -0,0 +1,626 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+from __future__ import annotations
+
+from typing import Any, Dict, List, Mapping, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import onnx
+import onnx.checker
+import onnx.defs
+import onnx.helper
+import onnx.shape_inference
+import torch
+from typing_extensions import TypeAlias
+
+import onnxscript
+from onnxscript import evaluator, onnx_opset
+from onnxscript import tensor as onnxscript_tensor
+from onnxscript._internal import param_manipulation, runtime_typing
+
+__all__ = [
+    "Graph",
+    "GraphEvaluator",
+    "GraphTensor",
+]
+
+
+ValidArgumentType: TypeAlias = Union[
+    "GraphTensor",
+    Sequence["GraphTensor"],
+    Sequence[float],
+    Sequence[int],
+    str,
+    int,
+    float,
+    bool,
+    None,
+]
+
+
+class GraphTensor(onnxscript_tensor.Tensor):
+    """An implementation of ONNX Tensors, based on a wrapper around numpy arrays.
+    Serves to define overloaded ops with an ONNX/ONNXScript semantics.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        onnx_type: str = "tensor_type",
+        opset=None,
+        torch_tensor: Optional[torch.Tensor] = None,
+    ):
+        super().__init__(None)
+        self._name = name
+        self._torch_tensor = torch_tensor
+        if torch_tensor is None:
+            self._shape = None
+            self._dtype = None
+        else:
+            self._dtype = torch_tensor.dtype
+            self._shape = tuple(
+                dim if isinstance(dim, int) else str(dim) for dim in torch_tensor.size()
+            )
+        # FIXME(justinhuby): Create a better way to determine the opset version
+        self._opset: Any = opset or onnx_opset.opset18
+
+        # TODO: Type class?
+        # NOTE: This is used to determine the type of the tensor
+        # when we convert it to onnx graph
+        if onnx_type not in {"tensor_type", "sequence_type", "optional_type"}:
```

Comment:
I am not sure about this. This sounds like an implicit coercion/cast of a tensor into a sequence of tensors or optional tensor. It may be simpler and better to make such cast explicit in the IR.

Reply:
Ok, on further looking at the code, it seems to me that this is meant to be a static representation of a runtime value. In that case, it would be better to rename the class (to something like `Value` or `Var`), and to _not_ inherit from Tensor.

Reply:
nit: the type-representation is incomplete. a sequence-type must describe the element-type too.

Reply:
We either need equivalent classes to represent these types (or must use strings like "sequence(tensor(int64))" as used in ONNX, but classes may be better).